subsys/dfu/dfu_multi_image/src/dfu_multi_image.c

```diff
@@ -0,0 +1,239 @@
+	res = res && zcbor_int32_decode(states, &image->id);
+	res = res && zcbor_tstr_expect_lit(states, "size");
+	res = res && zcbor_uint32_decode(states, &image->size);
+	res = res && zcbor_map_end_decode(states);
+
+	return res;
+}
+
+static int parse_header(void)
+{
+	if (ctx.cur_item_size == FIXED_HEADER_SIZE) {
+		/* Parse fixed header that contains the size of the following CBOR header */
+		ctx.cur_item_size += sys_get_le16(ctx.buffer);
+
+		return (ctx.cur_item_size <= ctx.buffer_size) ? 0 : -ENOMEM;
+	}
+
+	/* Parse CBOR header */
+	bool res;
+	uint_fast32_t image_count;
+	ZCBOR_STATE_D(states, CBOR_HEADER_NESTING_LEVEL, ctx.buffer + FIXED_HEADER_SIZE,
+		      ctx.cur_item_size - FIXED_HEADER_SIZE, 1);
+
+	res = zcbor_map_start_decode(states);
+	res = res && zcbor_tstr_expect_lit(states, "img");
+	res = res && zcbor_list_start_decode(states);
+	res = res && zcbor_multi_decode(1, CONFIG_DFU_MULTI_IMAGE_MAX_IMAGE_COUNT, &image_count,
+					(zcbor_decoder_t *)parse_image_info, states,
+					ctx.header.images, sizeof(struct image_info));
+	res = res && zcbor_list_end_decode(states);
+	res = res && zcbor_map_end_decode(states);
+
+	if (!res) {
+		return zcbor_pop_error(states);
+	}
+
+	ctx.header.image_count = image_count;
+	select_next_image();
+
+	return 0;
+}
+
+static const struct dfu_image_writer *current_image_writer(void)
+{
+	if (ctx.cur_image_no >= 0 && (size_t)ctx.cur_image_no < ctx.header.image_count) {
+		const int image_id = ctx.header.images[ctx.cur_image_no].id;
+
+		for (size_t i = 0; i < ctx.writer_count; i++) {
+			if (ctx.writers[i].image_id == image_id) {
+				return &ctx.writers[i];
+			}
+		}
+	}
+
+	return NULL;
+}
+
+static int process_current_item(const uint8_t *chunk, size_t chunk_size)
+{
+	int err = 0;
+
+	/* Process only remaining bytes of the current item (header or image) */
+	chunk_size = MIN(chunk_size, ctx.cur_item_size - ctx.cur_item_offset);
+
+	if (ctx.cur_image_no < 0) {
+		/* Parsing header */
+		memcpy(ctx.buffer + ctx.cur_item_offset, chunk, chunk_size);
+		ctx.cur_item_offset += chunk_size;
+
+		if (ctx.cur_item_offset == ctx.cur_item_size) {
+			err = parse_header();
+		}
+	} else {
+		/* Image data */
+		const struct dfu_image_writer *writer = current_image_writer();
+
+		if (!writer) {
+			err = -ESPIPE;
+		}
+
+		if (!err && ctx.cur_item_offset == 0) {
+			err = writer->open(writer->image_id);
+		}
+
+		if (!err) {
+			err = writer->write(chunk, chunk_size);
+			ctx.cur_item_offset += chunk_size;
+		}
+
+		if (ctx.cur_item_offset == ctx.cur_item_size) {
+			err = writer->close(true);
+			select_next_image();
+		}
+	}
+
+	if (err) {
+		return err;
+	}
+
+	return chunk_size;
+}
+
+int dfu_multi_image_init(uint8_t *buffer, size_t buffer_size, dfu_multi_image_seek_t seek)
+{
+	if (buffer == NULL || buffer_size < FIXED_HEADER_SIZE) {
+		return -EINVAL;
+	}
+
+	memset(&ctx, 0, sizeof(ctx));
+	ctx.buffer = buffer;
+	ctx.buffer_size = buffer_size;
+	ctx.seek = seek;
```

Comment:
In the current implementation the seek function is never called. Is this intentional?
If so, how the application should decide whether it should wait for the seek callback or simply move forward and download the next block?

Reply:
Yes, the only purpose I can see for it now is to skip S0/S1 images based on which MCUboot image is active. It's to be added once I or someone else decides to implement MCUboot upgrade and has a chance to test it. The application should never skip voluntarily, it should always wait for the seek callback. Currently, it will simply never call it, but if we agree on some scenarios in which we wish to do it, the API will be ready for that.

Reply:
The problem I have with skipping is that the library, which is meant to be pretty generic, would need to assume a specific image format, where the images are going to be written and what conditions should be applied to decide if an image is to be skipped. I think MCUboot upgrade is rather an obvious case, so perhaps we can sacrifice some generality here, but I don't have a good idea what to do with e.g. net core firmware upgrade. Us and HomeKit are not interested in that, anyway, but if you are and have an idea what API you would like for that, let me know.

Reply:
Implemented skipping using the mechanism which we discussed.